test/checks: improve AWS connection platform checks

benesch opened the pull request:
Various improvements to AWS connection testing:

  * Add end-to-end tests that run against real AWS for both credentials and assume role authentication, with both positive (expected to succeed) and negative (expected to fail) test cases.
  * Structure the validation error when a role fails to constrain the external ID, in order to give users more guidance on what's wrong and how to fix it.

### Motivation

Fixes #23055.

### Tips for reviewer

Go commit by commit. Details in commit messages.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

benesch commented:
@sploiselle @petrosagg @rjobanp — if y'all are working this week, can I ask one of you to take a look at this PR? Hoping to get it into this week's release so that we can put a bow on all the AWS connection work this year, teeing up the COPY TO/FROM S3 work for next year. 

Don't be scared off by the diffstat; it's almost entirely code movement and lockfile updates. Definitely go commit by commit. The only things that need close review are:

  1. The penultimate commit's addition of test/aws/mzcompose.py, which is a new E2E test for AWS connections. Everything else in the commit is `cargo vet` noise.
  2. The entirety of the final commit, which improves the `VALIDATE CONNECTION` error message.

benesch commented:
Nightlies were successful on the new AWS test: https://buildkite.com/materialize/nightlies/builds/5715

sploiselle reviewed (approved):
- nit isn't worth going back through CI
- Should we close out some of the TODOs in `aws-connection.td`? e.g. https://github.com/MaterializeInc/materialize/blob/main/test/aws/aws-connection/aws-connection.td#L65
- I don't have any special knowledge of the jump role, external ID scheme for AWS permission control. the tests are convincing but I haven't followed this work closely enough to know what to expect.

Approving but consider more of a +1 on the AWS integration test

benesch commented on src/storage-types/src/connections/aws.rs:
```diff
@@ -13,9 +13,12 @@ use anyhow::{anyhow, bail};
+use aws_sdk_sts::operation::get_caller_identity::GetCallerIdentityError;
 use aws_types::region::Region;
 use aws_types::SdkConfig;
 
```
Good eyes, thanks!

benesch commented on src/storage-types/src/connections/aws.rs:
```diff
@@ -386,3 +380,36 @@ impl AwsConnection {
+    pub fn hint(&self) -> Option<String> {
+        match self {
+            AwsConnectionValidationError::RoleDoesNotRequireExternalId { .. } => {
+                Some("See: https://materialize.com/s/aws-connection-role-trust-policy".into())
```
Yeah it's a shortlink added in this PR, so as long as I didn't typo (🤞🏽) it'll start working once I merge this.

benesch commented:
> * nit isn't worth going back through CI

Cleaned it up nonetheless!

> * Should we close out some of the TODOs in `aws-connection.td`? e.g. https://github.com/MaterializeInc/materialize/blob/main/test/aws/aws-connection/aws-connection.td#L65

Oh good spot, thanks. Replaced those TODOs with a pointer to /test/aws/.

> * I don't have any special knowledge of the jump role, external ID scheme for AWS permission control. the tests are convincing but I haven't followed this work closely enough to know what to expect.
> 
> Approving but consider more of a +1 on the AWS integration test

Ack. @moulimukherjee @guswynn — would appreciate a more in-depth post-merge look at this PR when you're back. Going to merge on @sploiselle's approval, though, since this is adding tests for a feature that's already merged so this strictly improves the situation.

moulimukherjee commented on test/aws/mzcompose.py:
```diff
@@ -0,0 +1,235 @@
+
+def test_credentials(c: Composition, ctx: TestContext):
+    # Create a user with an access key.
+    customer_user = f"testdrive-{ctx.seed}-Customer"
```
Should the users and roles created in this test be deleted as well as a cleanup after the test? If I am understanding right, these are actually created in Materialize Scratch account whenever they are run at least locally and we'll someday hit the limit?

<img width="682" alt="image" src="https://github.com/MaterializeInc/materialize/assets/2379129/6936f2b6-b28d-4a7a-86cb-e9bf13c26a24">
